cpp/include/cudf/detail/gather.cuh

```diff
@@ -255,7 +259,455 @@ struct column_gatherer_impl<dictionary32, MapItType> {
+ *
+ */
+#define LIST_GATHER_ITERATOR(__name, gather_data)                                                                      \
+  auto __name##span_upper_bound = thrust::make_transform_iterator(                                                     \
+    thrust::make_counting_iterator<size_type>(0),                                                                      \
+    [offsets = gather_data.offsets] __device__(size_type index) { return offsets[index + 1]; });                       \
+  column_view __name##base_offsets_v(*gather_data.base_offsets);                                                       \
+  size_type const* __name##base_offsets = __name##base_offsets_v.data<size_type>();                                    \
+  auto __name                           = thrust::make_transform_iterator(                                             \
+    thrust::make_counting_iterator<size_type>(0),                                            \
+    [span_upper_bound = __name##span_upper_bound,                                            \
+     offset_count     = __name##base_offsets_v.size(),                                       \
+     offsets          = gather_data.offsets,                                                 \
+     base_offsets     = __name##base_offsets] __device__(size_type index) {                      \
+      auto bound = thrust::upper_bound(                                                      \
+        thrust::device, span_upper_bound, span_upper_bound + offset_count, index);           \
+      size_type offset_index    = thrust::distance(span_upper_bound, bound);                 \
+      size_type offset_subindex = offset_index == 0 ? index : index - offsets[offset_index]; \
+      return offset_subindex + base_offsets[offset_index];                                   \
+    });
+
+  /**
+   * @brief Gather a leaf column from a hierarchy of list columns. The recursion
+   * terminates here.
+   *
+   * @param column View into the column to gather from
+   * @param gd The gather_data needed to construct a gather map iterator for this level
+   * @param nullify_out_of_bounds Nullify values in the gather map that are out of bounds
+   * @param mr Memory resource to use for all allocations
+   * @param stream CUDA stream on which to execute kernels
+   *
+   * @returns column with elements gathered based on `gather_data`
+   *
+   */
+  std::unique_ptr<column> gather_list_leaf(column_view const& column,
+                                           gather_data const& gd,
+                                           bool nullify_out_of_bounds,
+                                           rmm::mr::device_memory_resource* mr,
+                                           cudaStream_t stream)
+  {
+    // gather map for our child
+    LIST_GATHER_ITERATOR(child_gather_map, gd);
+
+    // size of the child gather map
+    size_type child_gather_map_size;
+    CUDA_TRY(cudaMemcpy(&child_gather_map_size,
```

Comment:
@davidwendt added a utility for this: https://github.com/rapidsai/cudf/blob/9aa0b984af6fbac29ac41aaf708b43544658f6a8/cpp/include/cudf/detail/get_value.cuh#L41